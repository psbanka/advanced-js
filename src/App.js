import React, {Component} from 'react'
// import './App.css'
import SearchBox from './SearchBox'

const SERVER_DATA = [
  {category: 'Sporting Goods', price: 49.99, stocked: true, name: 'Football'},
  {category: 'Sporting Goods', price: 9.99, stocked: true, name: 'Baseball'},
  {category: 'Sporting Goods', price: 29.99, stocked: false, name: 'Basketball'},
  {category: 'Electronics', price: 99.99, stocked: true, name: 'iPod Touch'},
  {category: 'Electronics', price: 399.99, stocked: false, name: 'iPhone 5'},
  {category: 'Electronics', price: 199.99, stocked: true, name: 'Nexus 7'}
]

class App extends Component {
  constructor (props) {
    super(props)
    this.state = {
      searchTerm: '',
      inStock: false,
      isBuying: {},
      total: 0
    }
  }

  /*
   * isBuying = {
   *   'Sporting GoodsBasketball': true,
   *   'ElectronicsNexus 7': true
   * }
   */

  onFilterTextInput (e) {
    this.setState({searchTerm: e.target.value})
  }

  onFilterCheckBoxInput (e) {
    this.setState({inStock: e.target.checked})
  }

  /**
   * a user has clicked a row
   * @param key {String} - combination of category and name to uniquely
   *   identify a row
   * @param value {bool} - selected or not
   * @param price {Number} - how much it costs
   */
  onIsBuying(key, value, price) {
    let newTotal
    if (value) {
      newTotal = this.state.total + price
    } else {
      newTotal = this.state.total - price
    }

    // Hey! if you want a variable as a key, put it in brackets!
    let newBuyObject = Object.assign(this.state.isBuying, {[key]: value})
    this.setState({isBuying: newBuyObject, total: newTotal})
  }

  render () {
    return (
      <div>
        <SearchBox
          searchTerm={this.state.searchTerm}
          inStock={this.state.inStock}
          onFilterTextInput={this.onFilterTextInput.bind(this)}
          onFilterCheckBoxInput={this.onFilterCheckBoxInput.bind(this)}
        />
        <ProductLine
          catalog={SERVER_DATA}
          searchTerm={this.state.searchTerm}
          inStock={this.state.inStock}
          isBuying={this.state.isBuying}
          onIsBuying={this.onIsBuying.bind(this)}
        />
        <p>{this.state.total}</p>
      </div>
    )
  }
}

class ProductLine extends Component {
  render () {
    return (
      <table>
        <ProductHeadline />
        <ProductData
          searchTerm={this.props.searchTerm}
          inStock={this.props.inStock}
          catalog={this.props.catalog}
          isBuying={this.props.isBuying}
          onIsBuying={this.props.onIsBuying}
        />
      </table>
    )
  }
}

class ProductHeadline extends Component {
  render () {
    return (
      <thead>
        <tr><td>Name</td><td>Price</td></tr>
      </thead>
    )
  }
}

class ProductData extends Component {

  _generateTableGuts () {
    let tableRow = []
    let categories = []
    this.props.catalog.forEach((catalogEntry) => {
      const newItem = categories.indexOf(catalogEntry.category) === -1
      if (newItem) {
        categories.push(catalogEntry.category)
      }
    })

    categories.forEach((currentCategory) => {
      tableRow.push(
        <tr key={currentCategory}>
          <td colSpan='2'><strong>{currentCategory}</strong></td>
        </tr>
      )
      this.props.catalog.forEach((catalogEntry) => {
        let key = `${currentCategory}${catalogEntry.name}`
        tableRow.push(
          <ProductRow
            key={key}
            {...catalogEntry}
            currentCategory={currentCategory}
            searchTerm={this.props.searchTerm}
            inStock={this.props.inStock}
            isBuying={this.props.isBuying}
            onIsBuying={this.props.onIsBuying}
          />
        )

      })
    })
    return tableRow
  }

  render () {
    const tableRow = this._generateTableGuts()
    return (
      <tbody>
        {tableRow}
      </tbody>
    )
  }
}

class ProductRow extends Component {

  handleOnIsBuying (e) {
    const value = e.target.checked
    let key = `${this.props.currentCategory}${this.props.name}`
    this.props.onIsBuying(key, value, this.props.price)
  }

  render () {
    const filterMatch = this.props.name.indexOf(this.props.searchTerm) !== -1
    let style = {color: 'black'}
    if (this.props.currentCategory === this.props.category) {
      if (!this.props.stocked) {
        style.color = 'red'
      }
      let key = `${this.props.currentCategory}${this.props.name}`
      let amIChecked = this.props.isBuying[key] || false
      if (!this.props.inStock || this.props.stocked) {
        if (filterMatch) {
          return(
            <tr>
              <td style={style}>
                <input
                  checked={amIChecked} type="checkbox"
                  onChange={this.handleOnIsBuying.bind(this)}
               />{this.props.name}
              </td>
              <td>${this.props.price}</td>
            </tr>
          )
        }
      }
    }
    return null
  }
}

export default App
