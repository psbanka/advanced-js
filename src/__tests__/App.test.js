import React from 'react'
import ReactDOM from 'react-dom'
import App from '../App'
import { shallow, mount, render } from 'enzyme'
import renderer from 'react-test-renderer';

/* global it describe */

describe('App', () => {
  it('does a snapshot check', () => {
    const component = renderer.create(<App />)
    const json = component.toJSON()
    expect(json).toMatchSnapshot()
  })

  describe('callbacks', () => {
    let wrapper, app

    beforeEach(() => {
      wrapper = shallow(<App/>)
      app = wrapper.instance()
    })

    it('performs onFilterTextInput', () => {
      app.onFilterTextInput({target: {value: 'cheese'}})
      expect(wrapper.state().searchTerm).toBe('cheese')
    })

    it('performs onFilterCheckboxInput', () => {
      app.onFilterCheckBoxInput({target: {checked: true}})
      expect(wrapper.state().inStock).toBe(true)
    })

  })

  describe('onIsBuying', () => {
    let app, wrapper

    beforeEach(() => {
      wrapper = shallow(<App/>)
      app = wrapper.instance()
    })

    it('properly increments price', () => {
      app.onIsBuying('product1', true, 299.99)
      const expected = {product1: true}
      expect(app.state.isBuying).toEqual(expected)
      expect(app.state.total).toEqual(299.99)
    })

    it('properly decrements price', () => {
      wrapper.setState({total: 299.99})
      app.onIsBuying('product1', false, 299.99)
      const expected = {product1: false}
      expect(app.state.isBuying).toEqual(expected)
      expect(app.state.total).toEqual(0)
    })
  })
})
