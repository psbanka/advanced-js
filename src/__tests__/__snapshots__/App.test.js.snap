// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App does a snapshot check 1`] = `
<div>
  <div
    className="SearchBox"
  >
    <button
      style={
        Object {
          "color": "red",
        }
      }
    >
      Confirm
    </button>
    <input
      id="in-stock-textbox"
      onChange={[Function]}
      placeholder="Search..."
      type="text"
      value=""
    />
    <p>
      <input
        checked={false}
        id="in-stock-checkbox"
        onClick={[Function]}
        type="checkbox"
      />
      Only show products in stock
    </p>
  </div>
  <table>
    <thead>
      <tr>
        <td>
          Name
        </td>
        <td>
          Price
        </td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          colSpan="2"
        >
          <strong>
            Sporting Goods
          </strong>
        </td>
      </tr>
      <tr
        className="product"
        style={
          Object {
            "color": "black",
          }
        }
      >
        <td>
          <input
            checked={false}
            id="sporting-goods-football"
            onChange={[Function]}
            type="checkbox"
          />
          Football
        </td>
        <td>
          $
          49.99
        </td>
      </tr>
      <tr
        className="product"
        style={
          Object {
            "color": "black",
          }
        }
      >
        <td>
          <input
            checked={false}
            id="sporting-goods-baseball"
            onChange={[Function]}
            type="checkbox"
          />
          Baseball
        </td>
        <td>
          $
          9.99
        </td>
      </tr>
      <tr
        className="product"
        style={
          Object {
            "color": "red",
          }
        }
      >
        <td>
          <input
            checked={false}
            id="sporting-goods-basketball"
            onChange={[Function]}
            type="checkbox"
          />
          Basketball
        </td>
        <td>
          $
          29.99
        </td>
      </tr>
      <tr>
        <td
          colSpan="2"
        >
          <strong>
            Electronics
          </strong>
        </td>
      </tr>
      <tr
        className="product"
        style={
          Object {
            "color": "black",
          }
        }
      >
        <td>
          <input
            checked={false}
            id="electronics-ipod-touch"
            onChange={[Function]}
            type="checkbox"
          />
          iPod Touch
        </td>
        <td>
          $
          99.99
        </td>
      </tr>
      <tr
        className="product"
        style={
          Object {
            "color": "red",
          }
        }
      >
        <td>
          <input
            checked={false}
            id="electronics-iphone-5"
            onChange={[Function]}
            type="checkbox"
          />
          iPhone 5
        </td>
        <td>
          $
          399.99
        </td>
      </tr>
      <tr
        className="product"
        style={
          Object {
            "color": "black",
          }
        }
      >
        <td>
          <input
            checked={false}
            id="electronics-nexus-7"
            onChange={[Function]}
            type="checkbox"
          />
          Nexus 7
        </td>
        <td>
          $
          199.99
        </td>
      </tr>
    </tbody>
  </table>
  <p
    id="total-box"
  >
    0
  </p>
</div>
`;
